Enable Config-File Based Control of EventPipe

brianrob opened the pull request:
With this change, EventPipe can be controlled without setting any environment variables.  To do so, do the following:

1. Create a config file of the form:

> ```
> OutputPath=/path/to/tracedirectory
> CircularMB=1024
> Providers=Microsoft-Windows-DotNETRuntime:0xFFFFFFFFFFFFFFFF:5,...
> ```

Only OutputPath is required.  All other variables are optional and will use default values.

2. Save the config file as EventPipeConfig.txt next to the entry assembly for the application.  This will start tracing (within 5 seconds, which is the polling interval).
3. When ready to stop tracing, delete EventPipeConfig.txt.  Within 1 second, tracing will be stopped and rundown will start.

When the process is complete a file with the name <entryassembly>.<processid>.netperf will be written to the location specified by the OutputPath config value.

The intention is that we will put this configuration behind a global tool and update the tool when the config-file based mechanism is replaced with something more long term such as a named pipe.

Note that this PR also brings back the ability to enable all events for diagnostic purposes.  This can be done by setting the following environment variables:

>```
>set COMPlus_EnableEventPipe=1
>set COMPlus_EventPipeConfig=*
>```

brianrob commented:
@dotnet-bot test this please

vancem commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipe.cs:
```diff
@@ -189,7 +199,7 @@ internal static class EventPipeInternal
         //
         [DllImport(JitHelpers.QCall, CharSet = CharSet.Unicode)]
-        internal static extern UInt64 Enable(string outputFile, uint circularBufferSizeInMB, long profilerSamplingRateInNanoseconds, EventPipeProviderConfiguration[] providers, int numProviders);
+        internal static extern UInt64 Enable(string outputFile, uint circularBufferSizeInMB, long profilerSamplingRateInNanoseconds, EventPipeProviderConfiguration[] providers, int numProviders, bool enableAllEvents);
```
I would really rather not have the 'All Events' special case wired into the 'Enable' API.   It is just not that valuable.   I was hoping instead that we could just treat '*' as a provider, and then have a special case that interprets it as a FILTER (set) rather than a particular provider (This also allows you extend this to general pattern matching (e.g Runtime*).

vancem commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeController.cs:
```diff
@@ -32,12 +34,19 @@ internal sealed class EventPipeController
         // Miscellaneous constants.
         private const string NetPerfFileExtension = ".netperf";
-        private const string MarkerFileExtension = ".ctl";
+        private const string ConfigFileName = "EventPipeConfig.txt";
```
Nit.  I am abivalent about *.txt.   Unix does not have suffixes, and this is not really a  text file anyway.   I am inclined to make it simply 'EventPipeConfig'.

vancem commented on src/System.Private.CoreLib/src/System/Diagnostics/Eventing/EventPipeController.cs:
```diff
@@ -144,7 +140,78 @@ private void PollForTracingCommand(object state)
+            return config;
+        }
+
+        private static EventPipeConfiguration BuildConfigFromEnvironment()
```
I would prefer that the File based and environment variable based modes are as similar as possible.   Thus we should not have EventPipeOutputFile, but EventPipeOuputPath, and it should be a directory (just like in the file case).   

I am ambivalent about specifying a directory and forcing the file to be named the same as the app.  (the less we mandate the better).  In the past, when I was also confronted with having to create potentially multiple output files, I chose asked the user to specify an OutputFilePrefix, which is a file path without a suffix, but it is expected that the system put the PROCESS_ID.netperf suffix after it.    What do you think of that?

vancem replied:
I am wondering if we should bother at all with environment variable control as we currently have it.   Here are three alternatives.
1. No environment variables, simply probe for the file on startup. 
2. Have one envionrment variable - COMPLUS_EventPipeConfigFile which indicates where the config file lives.  
3.  Have one environment variable COMPLUS_EventPipeConfiguration, which is the contents of the configuration file (and the file can have one more thing in it, which is the name of the configuration file itself (in our case AppBase\EventPipeConfig).    

All of these simplify things (more code reuse, or just plain simpler).   All allow for profiling at startup, but the first requires that you have access to the app, and requires modifying files where you may not wish to.  The first also requires another file touch (but given that we probed for *.json files, I would expect that to not be a problem). 

Thoughts?

brianrob commented:
@vancem, per our offline conversation, I have addressed your feedback:

1. Simplified the "enable all events" code to use look for a provider with the characteristics "*:0xFFFFFFFFFFFFFFFF:5".
2. Changed the name of the config file to app.eventpipeconfig.
3. Changed COMPlus_EventPipeOutputFile to COMPlus_EventPipeOutputPath.

NOTE: One thing that I did notice is that we're not able to get the entry assembly path if you specify COMPlus_EnableEventPipe=1 or have a config file present at start-up.  This is because we're too early in start-up.  In this case, and any other case where we can't get it, I have replaced it with the word "app", so that you get "app.pid.netperf".  I'm not concerned about the environment variable case, but I'm not super-thrilled with the pre-existing config file case.  At the same time, I like the idea of hooking in as early as we can, so I'm inclined to live with this at least for now.  We still have the PID disambiguator, but it is unfortunate if you want to write an automated tool that looks for trace files with a particular prefix.  Do you have any concerns about this?

anurse commented:
How would this interact with multi-instance applications? The config file would appear to affect all instances of a single physical binary on disk. That seems surprising to me as a user and generally not what I'd want an event collection tool to do by default. Usually I'm just interested in a single process. Is this not a scenario that we want to cover with this change?

vancem commented:
@anurse - Isolating the data collection to a single process before that process starts is actually a fundamentally problematic case.   You want to refer to one process, but it does not exist yet.   Environment variables don't do this because a process could spawn a sub-process (and indeed it could be the sub-process that you are interested in).   

A very simple way out of this is to adopt the 'app wide' control'  That is the commands go to all instances of the app, and you deal with it after the fact.   (thus the output is an output directory, and you logically are doing all apps (each of which have their own output file (hence the process ID in the output file name)).  You can always throw away any data you don't really want.      

Sure most of the time you are only interested in a single process.  But also most of the time there is only one instance of the app running.    ETW does this routinely (it includes EVERY process, not just this app) and it is fine.    

Note that it is reasonable to add a ProcessID=XXXX control to the EventPipeConfig which allows you to specify the process ID.   I dont' have a problem doing this, but keep in mind that that only works in the attach case (if you want to start a new process, you don't know how to specify XXXX).   

 Note that controling profiling with environment varibles is NOT usually convinient (often you don't know how your process got started, or how to inject new varibles into it)   Being able to step up to ANY app modify a file and get data without having to understand how to launch it is a very nice property.   We achieve this.   

We are open to suggestions, but this seems like a reasonable (and simple) design.   If you want the ProcessID=XXXX option we could add it now.

brianrob commented:
Adding a PID check is not hard, so if we do want to do this, now is the time.

brianrob commented:
I should also add that I am OK with adding such a thing, though as @vancem points out it only works in the attach case because you won't know the PID until it's too late.

vancem commented:
@brianrob  go ahead and add it.   It is clearly useful in the attach case and as you say, trivial to implement.

brianrob commented:
Sounds good.  I'll do that.

anurse commented:
> You want to refer to one process, but it does not exist yet.

Yes, I get that. However most of the scenarios we've looked at are "attach" scenarios where the process already exists and you want to avoid affecting other processes. Adding a ProcessId check sounds like a reasonable workaround for now.

I do think we need to talk a little more about the scenarios for this control mechanism as we develop it further. For example, it seems unfortunate that I need to know the path to the executable in order to enable tracing. Generally I'd rather be able to attach using only the process ID. However, that doesn't need to block this work, it can be handled as we look at further work here.

brianrob commented:
@vancem, the config file is now called <appName>.eventpipeconfig.  The previous issue was that Assembly.GetEntryAssembly() returned null, but I found the place where we set it, and have moved the callout to enable EventPipe right after this point.

vancem commented:
@brianrob  do you mean APPNAME.eventpipeconfig or do you literally mean .eventpipeconfig?   I am assuming it is the former... and the above was just a typo.

brianrob commented:
I mean APPNAME.eventpipeconfig where APPNAME == simple name of entry assembly.

brianrob commented:
@vancem, assuming tests pass, this should be ready to go.  Please let me know if you have any other questions or concerns.  Thanks.

vancem commented:
LGTM

brianrob commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu x64 Checked Innerloop Build and Test (Jit - TieredCompilation=0)
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test (Jit - TieredCompilation=0)
